server/src/Icepeak/Server/WebsocketServer/MultiSubscription.hs

```diff
@@ -0,0 +1,440 @@
+  , clientIsDirty :: MVar ()
+  , clientSubscriptions :: MVar (HashMap Path (MVar Value))
+  }
+
+doSubscribe :: Client -> [Path] -> IO ()
+doSubscribe client paths = do
+  let
+    core = clientCore client
+    conn = clientConn client
+
+    uuid = clientUuid client
+    isDirty = clientIsDirty client
+    subscriptions = clientSubscriptions client
+
+    coreClients = Core.coreClients core
+
+  pathsWithCurrentValue <- Monad.forM paths $
+    \path -> do
+      valueAtPath <- Core.getCurrentValue core path
+      pure (Text.intercalate "/" path, valueAtPath)
+
+  WS.sendTextData conn $
+    Aeson.encode $
+      ResponseSubscribeSuccess
+        { subscribeSuccessPathsValues = pathsWithCurrentValue
+        }
+
+  -- WARNING: Race condition, potentially miss an update when program is at this point
+
+  Monad.forM_ paths $ \newPath -> do
+    pathValueMVar <- MVar.newEmptyMVar
+
+    MVar.modifyMVar_
+      subscriptions
+      (pure . HashMap.insert newPath pathValueMVar)
+
+    MVar.modifyMVar_
+      coreClients
+      ( pure
+          . Subscription.subscribe
+            newPath
+            uuid
+            (\newValue -> do
+                Utils.writeToSub core pathValueMVar newValue
+                Monad.void $ MVar.tryPutMVar isDirty ())
+      )
+
+onPayloadSubscribeWithAuth
+  :: Client
+  -> JWT.VerifySigner
+  -> RequestSubscribe
+  -> IO ()
+onPayloadSubscribeWithAuth client secret (RequestSubscribe paths mbToken) = do
+  let conn = clientConn client
+  case mbToken of
+    Nothing -> do
+      WS.sendTextData conn $ -- 401  | No authorization token provided
+        Aeson.encode $
+          ResponseSubscribeFailure
+            { subscribeFailureStatusCode = 401
+            , subscribeFailureMessage = "No authorisation token provided"
+            , subscribeFailureExtraData = Nothing
+            , subscribeFailurePaths = Just paths
+            }
+    Just tokenBS -> do
+      let segmentedPaths = Text.splitOn "/" <$> paths :: [Path]
+      now <- Clock.getPOSIXTime
+      case JwtAuth.extractClaim now secret (Text.encodeUtf8 tokenBS) of
+        Left tokenError -> do
+          -- 403  | Authorization token was rejected / malformed |
+          WS.sendTextData conn $
+            Aeson.encode $
+              ResponseSubscribeFailure
+                { subscribeFailureStatusCode = 403
+                , subscribeFailureMessage = "Error while extracting claim from JWT: " <> Text.pack (show tokenError)
+                , subscribeFailureExtraData = Nothing
+                , subscribeFailurePaths = Just paths
+                }
+        Right authenticatedIcePeakClaim -> do
+          let
+            pathsIsAuth =
+              segmentedPaths
+                <&> ( \path ->
+                        ( path
+                        , Access.isAuthorizedByClaim authenticatedIcePeakClaim path Access.ModeRead
+                        )
+                    )
+            allAuth = and $ snd <$> pathsIsAuth
```

Comment:
Nit: this seems like a pretty-roundabout way of writing this, especially with the rather uncommon `<&>`. A bit simpler would be something like

```haskell
pathAuthorized path = Access.isAuthorizedByClaim authenticatedIcePeakClaim path Access.ModeRead
allAuth = all pathAutorized segmentedPaths
```

Reply:
It's roundabout because I use `allAuth` again to extract which paths failed to authorise and I send it to the client:
- https://github.com/channable/icepeak/blob/90909a7ed9c081d09c3fa8ebeab518043dda0fa4/server/src/Icepeak/Server/WebsocketServer/MultiSubscription.hs#L144-L146
But yeah probably best not to send it to the client?

Reply:
Ah, nevermind, I missed that. Time for another coffee :coffee:  All good then!